### PR TITLE
Add assert_unsafe_precondition to unchecked_{add,sub,neg,mul,shl,shr} methods

### PR DESCRIPTION
**Pull request by clarfonthey**
(Old PR is haunted, opening a new one. See #117494 for previous discussion.)

This ensures that these preconditions are actually checked in debug mode, and hopefully should let people know if they messed up. I've also replaced the calls (I could find) in the code that use these intrinsics directly with those that use these methods, so that the asserts actually apply.

More discussions on people misusing these methods in the tracking issue: https://github.com/rust-lang/rust/issues/85122.

### CODE REVIEW

**Comment by clarfonthey** [comment]
(Hastily added the commit I held off on and realised that tests are failing, so, I'll fix those soon and rebase as well.)

**Comment by saethlin** [comment]
Let's see if the ghosts followed us
@bors try @rust-timer queue

**Comment by clarfonthey** [comment]
So, I'm guessing that the answer is "about what we expected," although I'm not sure what percentage of those benchmarks are running under debug versus release mode and would have to do some more digging.

**Comment by saethlin** [comment]
Click on "comparison URL", that will take you to the page with all the useful data on it. Click around, explore.

I think the opt regressions might be addressed by https://github.com/rust-lang/rust/pull/121421, for sure that PR is supposed to eliminate the extra IR that would be generated.

The debug build regressions, well, keeping those down are why `assert_unsafe_precondition` internally stamps out a new function so that it can outline the check logic. It might be useful to investigate which check(s) are hottest using the technique I posted in the last PR. Certainly it will be educational to poke through IR, if you're into this sort of thing.

**Comment by clarfonthey** [comment]
I'm really inexperienced when it comes to debugging the compiler, so, I definitely do think that poking through the IR dumps would be cool, but I'm not sure how effective I'll be at it.

I think that this is blocked anyway on a couple optimisations anyway, so, at least I'll have time. I do still think that it'd be good to have a review that okays the code as-is regardless, even though we won't merge it until we get the perf situation sorted out, since that way we can be ready with the PR when that happens. I'll add a "don't merge" to the PR title in an effort to dissuade people from actually approving it before we sort out the blockers, but I don't want to actually add an S-blocked so folks still get around to the reviews.

**Comment by RalfJung** [comment]
Also carrying over this potential blocker from the other thread:

I think adding more of these checks needs to be blocked on [having a way to turn them off](https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Add.20a.20blanket.20flag.20to.20enable.2Fdisable.20code.E2.80.A6.20compiler-team.23625) separately from debug assertions, to avoid having more issues like [this](https://github.com/rust-lang/rust/issues/121245). (This is about runtime overhead, not compiletime overhead.) I love having these checks but I worry about the unintended side-effects of making even optimized debug-assertion builds a lot slower.

**Comment by saethlin** [comment]
r? saethlin

I'm basically reviewing/mentoring this PR already, and there would be a lot of history for a randomly-selected libs reviewer to go through.

**Comment by clarfonthey** [comment]
Going to mark this as blocked by #122520 since it seems clear that's going to be merged soon, and I'll have to fix the merge conflicts.

@rustbot blocked

**Comment by clarfonthey** [comment]
(note: this was purely to ensure that the changes are compatible; I'll properly rebase once the original is merged)

**Comment by clarfonthey** [comment]
Actually, probably worth considering this cc @saethlin to get a head-start before the other PR merges, but have the recent changes to how preconditions are done helped resolve the performance issues? At least miri will optimise these checks out now, but I'm not sure what will happen in other builds.

(It's fine if you don't get to this before the other PR merges, I just figured I'd bring it up now. The actual changes in the PR are in the last commit, since I already rebased over the other change in FCP.)

**Comment by saethlin** [comment]
What is "the other PR"?

Miri doesn't optimize by default, so maybe by optimization you were referring to this PR: https://github.com/rust-lang/rust/pull/122975#issuecomment-2025171546

**Comment by clarfonthey** [comment]
(properly rebased on master, and also removed the [don't merge] from the title since you're following well enough to only merge once the perf regression is fixed)

**Comment by clarfonthey** [comment]
Hmm, it is rather unfortunate that `ub_checks` creates two scopes instead of one, due to the addition of the const-runtime branch, but I'm not sure that this can be fixed, necessarily.

**Comment by saethlin** [comment]
It can be fixed by this strategy: https://github.com/rust-lang/rust/pull/121767

But extra scopes is not a problem in itself. The fact that the checks get weighted by the MIR inliner as a call even though when disabled they aren't lowered to a call, that's a problem.

**Comment by clarfonthey** [comment]
All good! Like I hopefully implied earlier, I'm in no rush to get this out, just wanted to prod a bit in case you were interested in reviewing sooner.

In terms of extra scopes, wasn't that an issue mentioned in an earlier comment, that even adding an extra scope to these methods can make compilation slower because these methods are called so often? Or am I just misremembering things? Either way, we could *probably* just leave the `const_eval_select` call without a wrapping function, since the result *will* be wrapped in a function either way once the MIR is generated.

**Comment by clarfonthey** [comment]
(actual conflict in #123226)

**Comment by clarfonthey** [comment]
Rebased and removed the two original concessions I had in this PR, namely:

* `unchecked_neg` now works for the pointer methods
* I'm calling the UB-checked version of `unchecked_sh[lr]` in `wrapping_sh[lr]` now

Seems that everything is working okay now, at least in the sense that I don't think the perf regressions would have gotten *worse* with recent changes.

File: library/core/src/num/int_macros.rs [hunk 1]
```diff
@@ -1165,7 +1204,7 @@ macro_rules! int_impl {
         #[rustc_const_stable(feature = "const_checked_int_methods", since = "1.47.0")]
         #[must_use = "this returns the result of the operation, \
                       without modifying the original"]
-        #[inline]
```
Comment (saethlin) [inline library/core/src/num/int_macros.rs:1168]: Why did you change this? I don't think we need to adjust the inline attribute on this and `checked_shr`, or at least we don't anymore. I know the codebase has shifted a bit over the course of this PR.

**Comment by saethlin** [comment]
I should have re-perf'd this when https://github.com/rust-lang/rust/pull/122975 landed. @clarfonthey can you resolve the merge conflicts? I'll start another perf run, then based on that I think the path forward for this is to experimentally determine if the compile time overhead is significantly reduced by adding `#[cfg(debug_assertions)]` to one of the `assert_unsafe_precondition!` calls.

File: library/core/src/num/int_macros.rs (hunk 1 above)
Comment (clarfonthey) [reply library/core/src/num/int_macros.rs:1168]: Yeah, I can just remove this.

**Comment by saethlin** [comment]
@clarfonthey After discussing this with the wg/team https://rust-lang.zulipchat.com/#narrow/stream/247081-t-compiler.2Fperformance/topic/How.20much.20do.2Fshould.20we.20care.20about.20optimized.20.2B.20incr.20builds.3F/near/438131519 I think the perf reports above do not block a merge.

If you fix up the merge conflicts, I think this is ready to go.

**Comment by clarfonthey** [comment]
Yeah uh, this is incredibly cursed. Suddenly adding these checks causes a panic that I have absolutely no idea how to debug:

<details><summary>`./x test library/core`</summary>
```
$ ./x test library/core
Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
Building stage0 library artifacts (x86_64-unknown-linux-gnu)
   Compiling core v0.0.0 (/home/ltdk/code/fork/rust/library/core)
   Compiling rustc-std-workspace-core v1.99.0 (/home/ltdk/code/fork/rust/library/rustc-std-workspace-core)
   Compiling compiler_builtins v0.1.109
   Compiling libc v0.2.153
   Compiling alloc v0.0.0 (/home/ltdk/code/fork/rust/library/alloc)
   Compiling cfg-if v1.0.0
   Compiling memchr v2.5.0
   Compiling adler v1.0.2
   Compiling rustc-demangle v0.1.24
   Compiling unwind v0.0.0 (/home/ltdk/code/fork/rust/library/unwind)
   Compiling rustc-std-workspace-alloc v1.99.0 (/home/ltdk/code/fork/rust/library/rustc-std-workspace-alloc)
   Compiling panic_abort v0.0.0 (/home/ltdk/code/fork/rust/library/panic_abort)
   Compiling panic_unwind v0.0.0 (/home/ltdk/code/fork/rust/library/panic_unwind)
   Compiling gimli v0.28.1
   Compiling miniz_oxide v0.7.2
   Compiling std_detect v0.1.5 (/home/ltdk/code/fork/rust/library/stdarch/crates/std_detect)
   Compiling object v0.32.2
   Compiling hashbrown v0.14.5
   Compiling addr2line v0.21.0
   Compiling std v0.0.0 (/home/ltdk/code/fork/rust/library/std)
   Compiling rustc-std-workspace-std v1.99.0 (/home/ltdk/code/fork/rust/library/rustc-std-workspace-std)
   Compiling proc_macro v0.0.0 (/home/ltdk/code/fork/rust/library/proc_macro)
   Compiling unicode-width v0.1.12
   Compiling getopts v0.2.21
   Compiling test v0.0.0 (/home/ltdk/code/fork/rust/library/test)
   Compiling sysroot v0.0.0 (/home/ltdk/code/fork/rust/library/sysroot)
    Finished `release` profile [optimized] target(s) in 13.13s
Testing  {core} (stage0 -> stage1, x86_64-unknown-linux-gnu)
   Compiling rand_core v0.6.4
   Compiling rand v0.8.5
   Compiling rand_xorshift v0.3.0
   Compiling core v0.0.0 (/home/ltdk/code/fork/rust/library/core)
thread 'rustc' panicked at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/compiler/rustc_middle/src/ty/layout.rs:770:21:
assertion failed: i < this.fields.count()
stack backtrace:
   0:     0x70c5ff751035 - std::backtrace_rs::backtrace::libunwind::trace::h97bf68a225bc049e
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/std/src/../../backtrace/src/backtrace/libunwind.rs:105:5
   1:     0x70c5ff751035 - std::backtrace_rs::backtrace::trace_unsynchronized::h43fd422959705afe
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x70c5ff751035 - std::sys_common::backtrace::_print_fmt::h78da944afe634dbd
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x70c5ff751035 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hdb567e5f04ec5928
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x70c5ff7a029b - core::fmt::rt::Argument::fmt::h97b972a553fcca94
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/core/src/fmt/rt.rs:165:63
   5:     0x70c5ff7a029b - core::fmt::write::he3ff590c9c56ce2e
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/core/src/fmt/mod.rs:1157:21
   6:     0x70c5ff745c5f - std::io::Write::write_fmt::hfe89b0aa813b7ceb
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/std/src/io/mod.rs:1832:15
   7:     0x70c5ff750e0e - std::sys_common::backtrace::_print::h493415ac3af32d28
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x70c5ff750e0e - std::sys_common::backtrace::print::hd0aef196e0c908d8
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x70c5ff753779 - std::panicking::default_hook::{{closure}}::h6216a28caae3e00c
  10:     0x70c5ff7534bd - std::panicking::default_hook::h5f0ed70b16f10e16
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/std/src/panicking.rs:298:9
  11:     0x70c5fc2e4077 - std[b95b9362c262c367]::panicking::update_hook::<alloc[5be86f0f543940e9]::boxed::Box<rustc_driver_impl[e2b437a5caec2333]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x70c5ff753e76 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hffefc6c0c7fdd8a5
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/alloc/src/boxed.rs:2036:9
  13:     0x70c5ff753e76 - std::panicking::rust_panic_with_hook::h3ca094bb5ec04c38
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/std/src/panicking.rs:799:13
  14:     0x70c5ff753beb - std::panicking::begin_panic_handler::{{closure}}::hbe1c2f230d27e9b3
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/std/src/panicking.rs:656:13
  15:     0x70c5ff7514f9 - std::sys_common::backtrace::__rust_end_short_backtrace::h7908055d1ea4f4f2
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x70c5ff753957 - rust_begin_unwind
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/std/src/panicking.rs:652:5
  17:     0x70c5ff79c763 - core::panicking::panic_fmt::h8ed2e06c66ae3e21
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/core/src/panicking.rs:72:14
  18:     0x70c5ff79c80c - core::panicking::panic::hb3f2ba737c36758e
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/core/src/panicking.rs:146:5
  19:     0x70c5fdbfa7fb - <rustc_middle[27c71549f86bc4e4]::ty::Ty as rustc_target[5f0bfc8513f43556]::abi::TyAbiInterface<_>>::ty_and_layout_field::field_ty_or_layout::<rustc_codegen_llvm[cce5530b50dc66bb]::context::CodegenCx>
  20:     0x70c5fe22013c - rustc_codegen_ssa[361d05578c83eb19]::mir::codegen_mir::<rustc_codegen_llvm[cce5530b50dc66bb]::builder::Builder>
  21:     0x70c5fe20d21e - rustc_codegen_llvm[cce5530b50dc66bb]::base::compile_codegen_unit::module_codegen
  22:     0x70c5fe209a81 - <rustc_codegen_llvm[cce5530b50dc66bb]::LlvmCodegenBackend as rustc_codegen_ssa[361d05578c83eb19]::traits::backend::ExtraBackendMethods>::compile_codegen_unit
  23:     0x70c5fe46c8e3 - <rustc_codegen_llvm[cce5530b50dc66bb]::LlvmCodegenBackend as rustc_codegen_ssa[361d05578c83eb19]::traits::backend::CodegenBackend>::codegen_crate
  24:     0x70c5fe448285 - rustc_interface[1c6f0d4c303def7d]::passes::start_codegen
  25:     0x70c5fe447a18 - <rustc_interface[1c6f0d4c303def7d]::queries::Queries>::codegen_and_build_linker
  26:     0x70c5fe1a92fe - rustc_interface[1c6f0d4c303def7d]::interface::run_compiler::<core[1795d7877ae73ac1]::result::Result<(), rustc_span[4cf4b8ad273124e]::ErrorGuaranteed>, rustc_driver_impl[e2b437a5caec2333]::run_compiler::{closure#0}>::{closure#1}
  27:     0x70c5fe19654b - std[b95b9362c262c367]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[1c6f0d4c303def7d]::util::run_in_thread_with_globals<rustc_interface[1c6f0d4c303def7d]::interface::run_compiler<core[1795d7877ae73ac1]::result::Result<(), rustc_span[4cf4b8ad273124e]::ErrorGuaranteed>, rustc_driver_impl[e2b437a5caec2333]::run_compiler::{closure#0}>::{closure#1}, core[1795d7877ae73ac1]::result::Result<(), rustc_span[4cf4b8ad273124e]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[1795d7877ae73ac1]::result::Result<(), rustc_span[4cf4b8ad273124e]::ErrorGuaranteed>>
  28:     0x70c5fe196346 - <<std[b95b9362c262c367]::thread::Builder>::spawn_unchecked_<rustc_interface[1c6f0d4c303def7d]::util::run_in_thread_with_globals<rustc_interface[1c6f0d4c303def7d]::interface::run_compiler<core[1795d7877ae73ac1]::result::Result<(), rustc_span[4cf4b8ad273124e]::ErrorGuaranteed>, rustc_driver_impl[e2b437a5caec2333]::run_compiler::{closure#0}>::{closure#1}, core[1795d7877ae73ac1]::result::Result<(), rustc_span[4cf4b8ad273124e]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[1795d7877ae73ac1]::result::Result<(), rustc_span[4cf4b8ad273124e]::ErrorGuaranteed>>::{closure#2} as core[1795d7877ae73ac1]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  29:     0x70c5ff75dcab - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h6f28929796384304
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/alloc/src/boxed.rs:2022:9
  30:     0x70c5ff75dcab - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hc19b559e890f4bf0
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/alloc/src/boxed.rs:2022:9
  31:     0x70c5ff75dcab - std::sys::pal::unix::thread::Thread::new::thread_start::h59a9f965326b1725
                               at /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/std/src/sys/pal/unix/thread.rs:108:17
  32:     0x70c5f92a955a - <unknown>
  33:     0x70c5f9326a5c - <unknown>
  34:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/home/ltdk/code/fork/rust/rustc-ice-2024-05-16T01_21_26-275469.txt` to your bug report

note: compiler flags: -C opt-level=3 -C embed-bitcode=no -Z unstable-options -C linker=gcc -C incremental=[REDACTED] -C symbol-mangling-version=legacy -Z unstable-options -Z unstable-options -Z macro-backtrace -C split-debuginfo=off -C prefer-dynamic -C llvm-args=-import-instr-limit=10 -Z inline-mir -Z inline-mir-preserve-debug -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -C lto=off -C force-frame-pointers=yes -Z crate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/") -Z binary-dep-depinfo -Z force-unstable-if-unmarked

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `core` (test "coretests")
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:00:27
```
</details>

**Comment by saethlin** [comment]
The ICE is present with
```rust
[rust]
optimize = 1
```
But goes away at
```toml
[rust]
optimize = 0
```
and doesn't seem affected by `-Zmir-opt-level=0`. Looks like an LLVM miscompile.

@DianQK We got another :upside_down_face: if you're willing to help out that would be awesome.

@clarfonthey Can you check if there was an LLVM major version bump landed between the test fail and the last time CI passed on this PR? I also know the library code has shifted around a bit, so if you're willing to do this kind of investigation it might help to identify which of the checks is relevant.

**Comment by DianQK** [comment]
> The ICE is present with
> 
> ```toml
> [rust]
> optimize = 1
> ```
> 
> But goes away at
> 
> ```toml
> [rust]
> optimize = 0
> ```
> 
> and doesn't seem affected by `-Zmir-opt-level=0`. Looks like an LLVM miscompile.
> 
> @DianQK We got another 🙃 if you're willing to help out that would be awesome.

Of course, I always prioritize addressing issues related to miscompile.

**Comment by clarfonthey** [comment]
Last passing commit appears to be 022942690d06807a3b820a64c9586bb067f2f0d0 whose in-tree parent is bd12986fd6659a3091cff7694b8225374f4a26fe; not 100% sure what I should be looking for in terms of an LLVM bump.

**Comment by clarfonthey** [comment]
Just doing a little bit of testing: removed all the checks from the code, then slowly started adding them back until it broke. Limited it down to the `unchecked_add`/`unchecked_sub`/`unchecked_neg` methods; removing the checks from these causes the error to disappear.

Gonna see if some of the code I changed to use these methods, instead of the intrinsics directly, might be the cause.

**Comment by clarfonthey** [comment]
…welp, it does appear that updating the pointer methods to use `unchecked_neg`, with checks, was the point of failure. I swear that `unchecked_neg` is cursed, since last time I was running into weird `rustc_allow_const_fn_unstable` bugs.

So, either removing the check from `unchecked_neg` or using the intrinsic directly for the library methods on pointers will avoid ICE. Specifically, the method where this is being applied is `sub`, which calls `self.offset((count as isize).unchecked_neg)`.

**Comment by clarfonthey** [comment]
Going to preemptively rebase on that commit so I can fix the merge conflicts. Thank you all for helping fix this!

File: library/core/src/slice/index.rs [hunk 2]
```diff
@@ -392,7 +391,7 @@ unsafe impl<T> SliceIndex<[T]> for ops::Range<usize> {
         );
         // SAFETY: see comments for `get_unchecked` above.
         unsafe {
-            let new_len = unchecked_sub(self.end, self.start);
+            let new_len = self.end.unchecked_sub(self.start);
```
[the remark this answers is not in the document]
Comment (clarfonthey) [reply library/core/src/slice/index.rs:394]: Fair, although we're ultimately going to have a bunch of checks duplicated when we have unsafe preconditions floating around *anyway*.

And besides, in a correct code base, the precondition checks will *always* be duplicated, since you should be only using them when they've been checked ahead of time anyway. It's just a matter of how many times they're duplicated.

File: library/core/src/num/int_macros.rs [hunk 3]
```diff
@@ -1111,9 +1141,18 @@ macro_rules! int_impl {
...
-            // SAFETY: the caller must uphold the safety contract for
-            // `unchecked_neg`.
-            unsafe { intrinsics::unchecked_sub(0, self) }
+            assert_unsafe_precondition!(
+                check_language_ub,
+                concat!(stringify!($SelfT), "::unchecked_neg cannot overflow"),
+                (
+                    lhs: $SelfT = self,
+                ) => !lhs.overflowing_neg().1,
+            );
```
[the remark this answers is not in the document]
Comment (clarfonthey) [reply library/core/src/num/int_macros.rs]: RIP, okay. Going to leave a proper comment alongside it to indicate why it's omitted for the bootstrap.

**Comment by clarfonthey** [comment]
Yup, already doing that and testing.

**Comment by clarfonthey** [comment]
(Worth adding that I marked it as blocked on the other PR, but it's ready for review if you want to go over the final MIR tests and verify that you're okay with the changes, assuming that there aren't any more PRs before this gets merged that affect the MIR output again.)

**Comment by clarfonthey** [comment]
(was trying out GitHub's sync branch UI; I'll do a proper rebase later)

**Comment by saethlin** [comment]
Hype hype hype
@bors try @rust-timer queue

**Comment by clarfonthey** [comment]
Just comparing to the last timer test, the one that was deemed "okay": https://github.com/rust-lang/rust/pull/121571#issuecomment-2045191209

current:

| mean | range | count |
| -- | -- | -- | -- |
| Regressions ❌  (primary) | 1.4% | [0.2%, 3.5%] | 30 |
| Regressions ❌  (secondary) | 1.1% | [0.3%, 3.7%] | 8 |
| Improvements ✅  (primary) | -1.0% | [-2.6%, -0.3%] | 4 |
| Improvements ✅  (secondary) | - | - | 0 |
| All ❌✅ (primary) | 1.1% | [-2.6%, 3.5%] | 34 |



old:

| mean | range | count |
| -- | -- | -- | -- |
| Regressions ❌  (primary) | 1.3% | [0.2%, 2.9%] | 31 |
| Regressions ❌  (secondary) | 1.8% | [0.3%, 3.7%] | 6 |
| Improvements ✅  (primary) | -0.4% | [-0.5%, -0.2%] | 4 |
| Improvements ✅  (secondary) | - | - | 0 |
| All ❌✅ (primary) | 1.2% | [-0.5%, 2.9%] | 35 |

**Comment by saethlin** [comment]
@bors r+
@rustbot label: +perf-regression-triaged The regressions in this PR are only in optimized + incremental builds, and are not huge. While we do care about build times in that configuration, we care a bit less than say debug + incr-patched. The additional checks at runtime that this PR brings are worth the small regressions to a single build mode. See https://rust-lang.zulipchat.com/#narrow/stream/247081-t-compiler.2Fperformance/topic/How.20much.20do.2Fshould.20we.20care.20about.20optimized.20.2B.20incr.20builds.3F for discussion of this.

**Comment by saethlin** [comment]
@bors r-
Your bless was incomplete, because you did it with `debug = true`. I still need to fix that footgun. Until I fix that, set `[rust] debug = false` in your `config.toml` then re-bless and you should see a diff.

**Comment by clarfonthey** [comment]
Ah, thank you for letting me know what exactly was happening; was a bit confused since I had done a full stage-1 test and it had still succeeded just fine.

**Comment by saethlin** [comment]
That looks right
@bors r+

**Comment by clarfonthey** [comment]
That *should* be the last test that needed to be fixed.

Realising now that the issue might be that I was only running stage1 tests, not stage2, but that's still kind of ridiculous that it only shows up on stage2.

**Comment by saethlin** [comment]
@bors r+

Hm, I'm not sure why that would be. I feel like I've run into this before but not recent enough to remember why...

**Comment by saethlin** [comment]
> I'll see what's actually being run in CI and emulate that instead.

In general, I run the test-various CI scripts as soon as anything bounces out of the queue.
```
./src/ci/docker/run.sh test-various
```
Or instead of `test-various`, any of the job names that your local Docker/Podman can execute.
https://rustc-dev-guide.rust-lang.org/tests/docker.html?highlight=docker#testing-with-docker

**Comment by clarfonthey** [comment]
That's a decent help; I hadn't realised that the CI scripts were that modular. For now though, I left stage2 running while I'm out doing errands and hope that will be sufficient; I'd prefer to be able to take advantage of cached builds if possible and not have to rerun everything every time.

**Comment by saethlin** [comment]
The CI scripts do all their work in `obj/` which means they will be cached between runs of the CI scripts, you just about double your disk usage if you're running them frequently.

I'm a bit wary that this new failure is target-specific or related to an interaction with a PR that's been merged in the meantime. Keep in mind that CI always tests the post-merge state, not a checkout of your branch. So for debugging strange CI failures I also try rebasing up my branch just in case.

**Comment by saethlin** [comment]
I rebased your branch over master and now I get this test failure:
```
diff -u --strip-trailing-cr "/checkout/obj/build/i686-unknown-linux-gnu/test/run-make/libtest-padding/libtest-padding"/bench.stdout bench.stdout
--- /checkout/obj/build/i686-unknown-linux-gnu/test/run-make/libtest-padding/libtest-padding/bench.stdout	2024-05-24 16:06:36.526391059 +0000
+++ bench.stdout	2024-05-12 04:30:42.484688349 +0000
@@ -2,7 +2,7 @@
 running 4 tests
 test short_test_name ... ignored
 test this_is_a_really_long_test_name ... ignored
-test short_bench_name                 ... bench:          ?? ns/iter (+/- ??)
+test short_bench_name                 ... bench:           ?? ns/iter (+/- ??)
 test this_is_a_really_long_bench_name ... bench:           ?? ns/iter (+/- ??)
 
 test result: ok. 0 passed; 0 failed; 2 ignored; 2 measured; 0 filtered out; finished in ??s
------------------------------------------
--- stderr -------------------------------
make: *** [Makefile:12: all] Error 1
------------------------------------------



failures:
    [run-make] tests/run-make/libtest-padding

test result: FAILED. 292 passed; 1 failed; 63 ignored; 0 measured; 0 filtered out; finished in 27.09s

Some tests failed in compiletest suite=run-make mode=run-make host=i686-unknown-linux-gnu target=i686-unknown-linux-gnu
Build completed unsuccessfully in 0:20:12
  local time: Fri May 24 16:06:55 UTC 2024
  network time: Fri, 24 May 2024 16:06:56 GMT
```
Note that this is the same test as failed in bors, but with a slightly different behavior.

**Comment by clarfonthey** [comment]
Yeah, that test seems to only be failing for me on the CI script, which is very weird. Will get back to you in five years once everything is done running and I've debugged things.

**Comment by saethlin** [comment]
I'm running a stage2 `optimize=0` build just in case that passes the test suite. `optimize=1` also failed on the same test.

**Comment by saethlin** [comment]
I applied this diff and I get exactly the same run-make libtest-padding test failure:
```diff
-ENV RUST_CONFIGURE_ARGS --build=i686-unknown-linux-gnu
+ENV RUSTFLAGS_NOT_BOOTSTRAP="-Zmir-opt-level=0"
+ENV RUST_CONFIGURE_ARGS \
+    --build=i686-unknown-linux-gnu \
+    --set rust.optimize=0
 # Skip some tests that are unlikely to be platform specific, to speed up
 # this slow job.
-ENV SCRIPT python3 ../x.py --stage 2 test \
-  --skip src/bootstrap \
-  --skip tests/rustdoc-js \
-  --skip src/tools/error_index_generator \
-  --skip src/tools/linkchecker
+ENV SCRIPT python3 ../x.py --stage 2 test run-make
```
```
diff -u --strip-trailing-cr "/checkout/obj/build/i686-unknown-linux-gnu/test/run-make/libtest-padding/libtest-padding"/bench.stdout bench.stdout
--- /checkout/obj/build/i686-unknown-linux-gnu/test/run-make/libtest-padding/libtest-padding/bench.stdout	2024-05-24 20:16:46.884851013 +0000
+++ bench.stdout	2024-05-12 04:30:42.484688349 +0000
@@ -2,8 +2,8 @@
 running 4 tests
 test short_test_name ... ignored
 test this_is_a_really_long_test_name ... ignored
-test short_bench_name                 ... bench:          ?? ns/iter (+/- ??)
-test this_is_a_really_long_bench_name ... bench:          ?? ns/iter (+/- ??)
+test short_bench_name                 ... bench:           ?? ns/iter (+/- ??)
+test this_is_a_really_long_bench_name ... bench:           ?? ns/iter (+/- ??)
```
So to me that indicates that this is not a buggy optimization, because I think I've turned them all off.

**Comment by saethlin** [comment]
I have more things to try, don't worry :)

**Comment by saethlin** [comment]
The problem was just a flaky test, and I would not have debugged it today without generous help from @jyn514.

@bors r+

**Comment by clarfonthey** [comment]
This PR just really doesn't want to be merged, I guess. Might have to wait until tomorrow if it's just some of the dependencies just not downloading.

**Comment by saethlin** [comment]
It looks like no PRs were merged for about 12 hours but one just got through.

@bors retry

**Comment by saethlin** [comment]
That is bizarre...

**Comment by saethlin** [comment]
Ah I see a number of recent PRs are bouncing off this. https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/Numerous.20errors.20downloading.20mingw/near/440643904

**Comment by saethlin** [comment]
The last 2 PRs have landed without error :crossed_fingers: 
@bors r+

**Comment by clarfonthey** [comment]
Thank you to everyone for helping get this merged!
